File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/net/ClientPacketHandler.java
package pize.tests.voxelgame.client.net;

import pize.Pize;
import pize.math.Maths;
import pize.math.vecmath.vector.Vec3f;
import pize.net.tcp.TcpConnection;
import pize.net.tcp.TcpListener;
import pize.net.tcp.packet.PacketHandler;
import pize.net.tcp.packet.PacketInfo;
import pize.net.tcp.packet.Packets;
import pize.tests.voxelgame.client.ClientGame;
import pize.tests.voxelgame.client.entity.LocalPlayer;
import pize.tests.voxelgame.client.entity.RemotePlayer;
import pize.tests.voxelgame.main.chat.MessageSourceServer;
import pize.tests.voxelgame.main.entity.Entity;
import pize.tests.voxelgame.main.net.packet.*;
import pize.tests.voxelgame.main.text.Component;

public class ClientPacketHandler implements TcpListener, PacketHandler{
    
    private final ClientGame game;
    
    public ClientPacketHandler(ClientGame game){
        this.game = game;
    }
    
    public ClientGame getGame(){
        return game;
    }
    
    
    public static int rx;
    public static int rxCounter;
    
    
    @Override
    public void received(byte[] bytes, TcpConnection sender){
        final PacketInfo packetInfo = Packets.getPacketInfo(bytes);
        if(packetInfo == null)
            return;
        rxCounter++;
        switch(packetInfo.getPacketID()){
            
            // Login
            case CBPacketDisconnect.PACKET_ID ->{
                final CBPacketDisconnect packet = packetInfo.readPacket(new CBPacketDisconnect());
                
                game.disconnect();
                System.out.println("[Client]: Connection closed: " + packet.reasonComponent);
            }
            
            case CBPacketEncryptStart.PACKET_ID ->{
                final CBPacketEncryptStart packet = packetInfo.readPacket(new CBPacketEncryptStart());
                
                byte[] encryptedClientKey = packet.publicServerKey.encrypt(game.getEncryptKey().getKey().getEncoded());
                new SBPacketEncryptEnd(encryptedClientKey).write(sender);
                
                sender.encode(game.getEncryptKey());// * шифрование *
                
                new SBPacketAuth(game.getSession().getSessionToken()).write(sender);
            }
            
            // Game
            case CBPacketSpawnInfo.PACKET_ID ->{
                final CBPacketSpawnInfo packet = packetInfo.readPacket(new CBPacketSpawnInfo());
                
                game.createClientLevel(packet.levelName);
                game.spawnPlayer(packet.position);
                game.getLevel().getChunkManager().startLoadChunks();
                
                game.sendPacket(new SBPacketRenderDistance(game.getSession().getOptions().getRenderDistance()));
            }

            case CBPacketPlaySound.PACKET_ID -> {
                final CBPacketPlaySound packet = packetInfo.readPacket(new CBPacketPlaySound());

                final Vec3f camPosition = game.getCamera().position;

                Pize.execSync(() ->
                    game.getSession().getAudioPlayer().play(
                        packet.soundID,
                        packet.volume, packet.pitch,
                        packet.x - camPosition.x,
                        packet.y - camPosition.y,
                        packet.z - camPosition.z
                    )
                );
            }
            
            case CBPacketTime.PACKET_ID -> {
                final CBPacketTime packet = packetInfo.readPacket(new CBPacketTime());

                game.getTime().setTicks(packet.gameTimeTicks);
            }
            
            case CBPacketAbilities.PACKET_ID -> {
                final CBPacketAbilities packet = packetInfo.readPacket(new CBPacketAbilities());

                game.getPlayer().setFlyEnabled(packet.flyEnabled);
            }
            
            case CBPacketTeleportPlayer.PACKET_ID ->{
                final CBPacketTeleportPlayer packet = packetInfo.readPacket(new CBPacketTeleportPlayer());
                
                final LocalPlayer localPlayer = game.getPlayer();
                if(localPlayer == null)
                    break;
                
                // Load another level
                if(!packet.levelName.equals(localPlayer.getLevel().getConfiguration().getName())){
                    game.createClientLevel(packet.levelName);
                    game.getLevel().getChunkManager().startLoadChunks();
                    localPlayer.setLevel(game.getLevel());
                }
                
                localPlayer.getPosition().set(packet.position);
                localPlayer.getRotation().set(packet.rotation);
            }
            
            case CBPacketChatMessage.PACKET_ID ->{
                final CBPacketChatMessage packet = packetInfo.readPacket(new CBPacketChatMessage());
                game.getChat().putMessage(packet);
            }
            
            case CBPacketPlayerSneaking.PACKET_ID ->{
                final CBPacketPlayerSneaking packet = packetInfo.readPacket(new CBPacketPlayerSneaking());
                
                final Entity targetEntity = game.getLevel().getEntity(packet.playerUUID);
                if(targetEntity instanceof RemotePlayer player)
                    player.setSneaking(packet.sneaking);
            }
            
            case CBPacketEntityMove.PACKET_ID ->{
                final CBPacketEntityMove packet = packetInfo.readPacket(new CBPacketEntityMove());

                Entity targetEntity = game.getLevel().getEntity(packet.uuid);
                if(targetEntity == null && game.getPlayer().getUUID() == packet.uuid)
                    targetEntity = game.getPlayer();
                
                if(targetEntity != null){
                    targetEntity.getPosition().set(packet.position);
                    targetEntity.getRotation().set(packet.rotation);
                    targetEntity.getVelocity().set(packet.velocity);
                }
            }

            case CBPacketSpawnEntity.PACKET_ID ->{
                final CBPacketSpawnEntity packet = packetInfo.readPacket(new CBPacketSpawnEntity());

                final Entity entity = packet.type.createEntity(game.getLevel());
                entity.getPosition().set(packet.position);
                entity.getRotation().set(packet.rotation);
                entity.setUUID(packet.uuid);

                game.getLevel().addEntity(entity);
            }
            
            case CBPacketRemoveEntity.PACKET_ID ->{
                final CBPacketRemoveEntity packet = packetInfo.readPacket(new CBPacketRemoveEntity());
                
                game.getLevel().removeEntity(packet.uuid);
            }
            
            case CBPacketSpawnPlayer.PACKET_ID ->{
                final CBPacketSpawnPlayer packet = packetInfo.readPacket(new CBPacketSpawnPlayer());
                
                final RemotePlayer remotePlayer = new RemotePlayer(game.getLevel(), packet.playerName);
                remotePlayer.getPosition().set(packet.position);
                remotePlayer.getRotation().set(packet.rotation);
                remotePlayer.setUUID(packet.uuid);
                
                game.getLevel().addEntity(remotePlayer);
            }
            
            case CBPacketBlockUpdate.PACKET_ID ->{
                final CBPacketBlockUpdate packet = packetInfo.readPacket(new CBPacketBlockUpdate());
                game.getLevel().setBlock(packet.x, packet.y, packet.z, packet.state);
            }
            
            case CBPacketChunk.PACKET_ID ->{
                final CBPacketChunk packet = packetInfo.readPacket(new CBPacketChunk());
                game.getLevel().getChunkManager().receivedChunk(packet);
            }
            
            // Ping
            case CBPacketPong.PACKET_ID -> {
                final CBPacketPong packet = packetInfo.readPacket(new CBPacketPong());
                
                final String message = "Ping - " + String.format("%.5f", (System.nanoTime() - packet.timeNanos) / Maths.NanosInSecond) + " ms";
                game.getChat().putMessage(new MessageSourceServer(), new Component().text(message));
                System.out.println("[Client]: " + message);
            }
            
        }
    }
    
    @Override
    public void connected(TcpConnection connection){ }
    
    @Override
    public void disconnected(TcpConnection connection){
        Pize.exit();
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/net/ClientPacketHandler.kt
package pize.tests.voxelgame.client.net

import pize.Pize.execSync
import pize.Pize.exit
import pize.graphics.texture.Region.set
import pize.graphics.util.color.Color.set
import pize.math.Maths
import pize.math.util.EulerAngles.set
import pize.math.vecmath.matrix.Matrix4f.set
import pize.math.vecmath.vector.Vec2f.set
import pize.math.vecmath.vector.Vec3f.set
import pize.math.vecmath.vector.Vec3i.set
import pize.net.tcp.TcpConnection
import pize.net.tcp.TcpListener
import pize.net.tcp.packet.PacketHandler
import pize.net.tcp.packet.Packets.getPacketInfo
import pize.tests.voxelgame.client.ClientGame
import pize.tests.voxelgame.client.entity.RemotePlayer
import pize.tests.voxelgame.main.chat.MessageSourceServer
import pize.tests.voxelgame.main.net.packet.*
import pize.tests.voxelgame.main.text.Component

class ClientPacketHandler(val game: ClientGame) : TcpListener, PacketHandler {

    override fun received(bytes: ByteArray?, sender: TcpConnection?) {
        val packetInfo = getPacketInfo(bytes!!) ?: return
        rxCounter++
        when (packetInfo.packetID) {
            CBPacketDisconnect.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketDisconnect())
                game.disconnect()
                println("[Client]: Connection closed: " + packet.reasonComponent)
            }

            CBPacketEncryptStart.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketEncryptStart())
                val encryptedClientKey = packet.publicServerKey!!.encrypt(game.encryptKey.key.getEncoded())
                SBPacketEncryptEnd(encryptedClientKey).write(sender)
                sender!!.encode(game.encryptKey) // * шифрование *
                SBPacketAuth(game.session.sessionToken).write(sender)
            }

            CBPacketSpawnInfo.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketSpawnInfo())
                game.createClientLevel(packet.levelName)
                game.spawnPlayer(packet.position)
                game.level.chunkManager.startLoadChunks()
                game.sendPacket(SBPacketRenderDistance(game.session.options.renderDistance))
            }

            CBPacketPlaySound.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketPlaySound())
                val camPosition = game.camera.position
                execSync {
                    game.session.audioPlayer.play(
                        packet.soundID,
                        packet.volume, packet.pitch,
                        packet.x - camPosition.x,
                        packet.y - camPosition.y,
                        packet.z - camPosition.z
                    )
                }
            }

            CBPacketTime.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketTime())
                game.time.ticks = packet.gameTimeTicks
            }

            CBPacketAbilities.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketAbilities())
                game.player.isFlyEnabled = packet.flyEnabled
            }

            CBPacketTeleportPlayer.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketTeleportPlayer())
                val localPlayer = game.player ?: break

                // Load another level
                if (packet.levelName != localPlayer.level.configuration.name) {
                    game.createClientLevel(packet.levelName)
                    game.level.chunkManager.startLoadChunks()
                    localPlayer.setLevel(game.level)
                }
                localPlayer.position.set(packet.position!!)
                localPlayer.rotation.set(packet.rotation)
            }

            CBPacketChatMessage.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketChatMessage())
                game.chat.putMessage(packet)
            }

            CBPacketPlayerSneaking.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketPlayerSneaking())
                val targetEntity = game.level.getEntity(packet.playerUUID)
                (targetEntity as? RemotePlayer)?.setSneaking(packet.sneaking)
            }

            CBPacketEntityMove.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketEntityMove())
                var targetEntity = game.level.getEntity(packet.uuid)
                if (targetEntity == null && game.player.uuid === packet.uuid) targetEntity = game.player
                if (targetEntity != null) {
                    targetEntity.position.set(packet.position!!)
                    targetEntity.rotation.set(packet.rotation)
                    targetEntity.velocity.set(packet.velocity)
                }
            }

            CBPacketSpawnEntity.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketSpawnEntity())
                val entity = packet.type!!.createEntity(game.level)
                entity!!.position.set(packet.position!!)
                entity.rotation.set(packet.rotation)
                entity.uuid = packet.uuid
                game.level.addEntity(entity)
            }

            CBPacketRemoveEntity.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketRemoveEntity())
                game.level.removeEntity(packet.uuid)
            }

            CBPacketSpawnPlayer.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketSpawnPlayer())
                val remotePlayer = RemotePlayer(game.level, packet.playerName)
                remotePlayer.position.set(packet.position!!)
                remotePlayer.rotation.set(packet.rotation)
                remotePlayer.uuid = packet.uuid
                game.level.addEntity(remotePlayer)
            }

            CBPacketBlockUpdate.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketBlockUpdate())
                game.level.setBlock(packet.x, packet.y, packet.z, packet.state)
            }

            CBPacketChunk.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketChunk())
                game.level.chunkManager.receivedChunk(packet)
            }

            CBPacketPong.Companion.PACKET_ID -> {
                val packet = packetInfo.readPacket(CBPacketPong())
                val message = "Ping - " + String.format(
                    "%.5f",
                    (System.nanoTime() - packet.timeNanos) / Maths.NanosInSecond
                ) + " ms"
                game.chat.putMessage(MessageSourceServer(), Component().text(message))
                println("[Client]: $message")
            }
        }
    }

    override fun connected(connection: TcpConnection?) {}
    override fun disconnected(connection: TcpConnection?) {
        exit()
    }

    companion object {
        var rx = 0
        var rxCounter = 0
    }
}
